ipam/mas.go

```diff
@@ -5,129 +5,193 @@ package ipam
+		return err
+	}
+
+	err = populateAddressSpace(local, sdnInterfaces, localInterfaces)
+	if err != nil {
+		return err
+	}
 
-	// Decode JSON object.
-	var obj jsonObject
-	decoder := json.NewDecoder(resp.Body)
-	err = decoder.Decode(&obj)
+	// Set the local address space as active.
+	err = source.sink.setAddressSpace(local)
 	if err != nil {
 		return err
 	}
 
-	// Add the IP addresses to the local address space.
-	for _, v := range obj.IPs {
-		address := net.ParseIP(v.IP)
-		subnet := net.IPNet{
-			IP:   net.ParseIP(v.IP),
-			Mask: net.IPMask(net.ParseIP(v.Mask)),
+	source.fileLoaded = true
+
+	return nil
+}
+
+func getSDNInterfaces(fileLocation string) (*NetworkInterfaces, error) {
+	data, err := ioutil.ReadFile(fileLocation)
+	if err != nil {
+		return nil, err
+	}
+
+	interfaces := &NetworkInterfaces{}
+	err = json.Unmarshal(data, interfaces)
+	if err != nil {
+		return nil, err
+	}
+
+	return interfaces, nil
+}
+
+func populateAddressSpace(localAddressSpace *addressSpace, sdnInterfaces *NetworkInterfaces, localInterfaces []net.Interface) error {
+
+	//Find the interface with matching MacAddress or Name
+	for _, sdnIf := range sdnInterfaces.Interfaces {
+		ifName := ""
+
+		for _, localIf := range localInterfaces {
+			if macAddressesEqual(sdnIf.MacAddress, localIf.HardwareAddr.String()) {
+				ifName = localIf.Name
+				break
+			}
 		}
 
-		ap, err := local.newAddressPool("eth0", 0, &subnet)
-		if err != nil {
-			log.Printf("[ipam] Failed to create pool:%v err:%v.", subnet, err)
+		// Skip if interface is not found.
+		if ifName == "" {
+			log.Printf("[ipam] Failed to find interface with MAC address:%v", sdnIf.MacAddress)
 			continue
 		}
 
-		_, err = ap.newAddressRecord(&address)
-		if err != nil {
-			log.Printf("[ipam] Failed to create address:%v err:%v.", address, err)
-			continue
+		// Prioritize secondary interfaces.
+		priority := 0
+		if !sdnIf.IsPrimary {
+			priority = 1
 		}
-	}
 
-	// Set the local address space as active.
-	s.sink.setAddressSpace(local)
+		for _, subnet := range sdnIf.IPSubnets {
+			_, network, err := net.ParseCIDR(subnet.Prefix)
+			if err != nil {
+				log.Printf("[ipam] Failed to parse subnet:%v err:%v.", subnet.Prefix, err)
+				continue
+			}
+
+			addressPool, err := localAddressSpace.newAddressPool(ifName, priority, network)
+			if err != nil {
+				log.Printf("[ipam] Failed to create pool:%v ifName:%v err:%v.", subnet, ifName, err)
+				continue
+			}
+
+			// Add the IP addresses to the localAddressSpace address space.
+			for _, ipAddr := range subnet.IPAddresses {
+				// Primary addresses are reserved for the host.
+				if ipAddr.IsPrimary {
+					continue
+				}
+
+				address := net.ParseIP(ipAddr.Address)
+
+				_, err = addressPool.newAddressRecord(&address)
+				if err != nil {
+					log.Printf("[ipam] Failed to create address:%v err:%v.", address, err)
+					continue
+				}
+			}
+		}
+	}
 
 	return nil
 }
+
+func macAddressesEqual(macAddress1 string, macAddress2 string) bool {
```

Comment:
If the mac from NetworkInterfaces is already formatted, you can either ParseMAC and compare with HardwareAddr or compare directly with HardwareAddr since you've already "String()"'d it.

Reply:
The problem is that the `ParseMAC()` function expects the address to be in the form `00:0d:3a:6e:18:25` and the `String()` function also prints it out that way. Wireserver returns the address in the form `000D3A6E1825` and the file we generate will be the same. `ParseMAC()` fails on this type of formatting.